csrc/preseg_passes/consecutive_cast.cpp

```diff
@@ -92,92 +208,165 @@ Val* replaceInputInCast(Val* cast_output, Val* new_input) {
 //
 //        b. otherwise, we can't bypass `lo_anchor` cast, we rewire this
 //        section as `starting_anchor`->`lo_anchor`->`expr->output(0)`
+Expr* removeChainedCasts(Expr* expr, std::unordered_set<Expr*>& visited) {
+  std::list<Val*> chain_cast_vals;
+  auto prev_expr = expr->input(0)->definition();
+  while (isCast(prev_expr)) {
```

Comment:
Nit: because `removedChainedCasts` is called in a while loop, we don't really need another while loop here. That may help simplify the logic.

Reply:
I don't see an easy way to simplify the logic.
Since the logic here [relies on analyzing the chain of casts and tries to simplify it based on the high/low water mark of cast types](https://github.com/NVIDIA/Fuser/blob/f82ee8ee142cbe029b7f85cac793b99ea7094360/csrc/preseg_passes/consecutive_cast.cpp#L174-L210)

We can embed that into the parent while loop, but I think that's going to make it harder to read, since we are interleaving it with the meta/cast swap part.

Reply:
I believe we can fold iteratively without having to analyze the whole chain. We only need to apply two rules to a fixed point. 
1. when we see `dtype A -> (cast) -> dtype B -> (cast) -> dtype C` and that B includes A, simplify it to `dtype A -> (cast) -> dtype C`. 
2. when we see `dtype A -> (cast) -> dtype A` (i.e. the producer TV and consumer have the same dtype), remove the cast. 
Am I understanding this correctly? 

(In case I wasn't clear, I'm not requesting changes in this PR)

Reply:
That's a fair thing to try.

The existing approach is a bit more aggressive, but it might not be necessary to have that and it's a fair ask to change it for simpler implementation. (Also, I realized that using `isInclusiveType` to figure out the lowest precision here might not be numerically safe.)

I'll follow up with a quick refactor on that. Need to double check then if any existing tests need to be modified.